apis/Google.LongRunning/Google.LongRunning/Operation.cs

```diff
@@ -130,16 +139,32 @@ public T Result
         /// <param name="pollSettings">The settings to use for repeated polling, or null
         /// to use the default poll settings (poll once every 10 seconds, forever).</param>
         /// <param name="callSettings">The call settings to apply on each call, or null for default settings.</param>
+        /// <param name="metadataCallback">The callback to invoke whenever a polled result contains metadata, or null
+        /// for no callback.</param>
         /// <returns>The completed operation, which can then be checked for errors or a result.</returns>
-        public Operation<T> PollUntilCompleted(PollSettings pollSettings = null, CallSettings callSettings = null)
+        public Operation<TResponse, TMetadata> PollUntilCompleted(
+            PollSettings pollSettings = null,
+            CallSettings callSettings = null,
+            Action<TMetadata> metadataCallback = null)
         {
             if (IsCompleted)
             {
                 return this;
             }
             callSettings = Client.GetEffectiveCallSettingsForGetOperation(callSettings);
 
-            Func<DateTime?, Operation<T>> pollAction = deadline => PollOnce(callSettings.WithEarlierDeadline(deadline, Client.Clock));
+            Func<DateTime?, Operation<TResponse, TMetadata>> pollAction =
+                deadline =>
+                {
+                    var result = PollOnce(callSettings.WithEarlierDeadline(deadline, Client.Clock));
+                    TMetadata metadata;
+                    if (metadataCallback != null && (metadata = result.Metadata) != null)
```

Comment:
I wasn't either - it avoids nested if statements though, which also aren't terribly nice. Happy to go that way if you'd prefer though.